app/src/sensors.c

```diff
@@ -18,65 +18,144 @@ LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 
 #if ZMK_KEYMAP_HAS_SENSORS
 
-struct sensors_data_item {
-    uint8_t sensor_number;
+struct sensors_item_cfg {
+    uint8_t sensor_index;
+    const struct zmk_sensor_config *config;
     const struct device *dev;
     struct sensor_trigger trigger;
 };
 
-#define _SENSOR_ITEM(node)                                                                         \
+#define _SENSOR_ITEM(idx, node)                                                                    \
     {                                                                                              \
-        .dev = NULL, .trigger = {.type = SENSOR_TRIG_DELTA, .chan = SENSOR_CHAN_ROTATION }         \
+        .dev = DEVICE_DT_GET_OR_NULL(node),                                                        \
+        .trigger = {.type = SENSOR_TRIG_DATA_READY, .chan = SENSOR_CHAN_ROTATION},                 \
+        .config = &configs[idx]                                                                    \
     }
+#define SENSOR_ITEM(idx, _i) _SENSOR_ITEM(idx, ZMK_KEYMAP_SENSORS_BY_IDX(idx))
 
-#define SENSOR_ITEM(idx, _node)                                                                    \
-    COND_CODE_1(DT_NODE_HAS_STATUS(ZMK_KEYMAP_SENSORS_BY_IDX(idx), okay),                          \
-                (_SENSOR_ITEM(ZMK_KEYMAP_SENSORS_BY_IDX(idx))), ({}))
+#define PLUS_ONE(n) +1
+#define ZMK_KEYMAP_SENSORS_CHILD_COUNT (0 DT_FOREACH_CHILD(ZMK_KEYMAP_SENSORS_NODE, PLUS_ONE))
+static struct zmk_sensor_config configs[] = {
 
-static struct sensors_data_item sensors[] = {LISTIFY(ZMK_KEYMAP_SENSORS_LEN, SENSOR_ITEM, (, ), 0)};
+#if ZMK_KEYMAP_SENSORS_CHILD_COUNT > 0
 
-static void zmk_sensors_trigger_handler(const struct device *dev,
-                                        const struct sensor_trigger *trigger) {
-    int err;
-    const struct sensors_data_item *item = CONTAINER_OF(trigger, struct sensors_data_item, trigger);
+#define SENSOR_CHILD_ITEM(node)                                                                    \
```

Comment:
Minor: `#define`ing something in the middle of an array initializer is a bit confusing. Would be a bit easier to read if this was

```c
#define SENSOR_CHILD_ITEM ...
#define SENSOR_CHILD_DEFAULTS ...

static struct zmk_sensor_config configs[] = {
#if ZMK_KEYMAP_SENSORS_CHILD_COUNT > 0
    DT_FOREACH_CHILD_SEP(ZMK_KEYMAP_SENSORS_NODE, SENSOR_CHILD_ITEM, (, ))
#else
    LISTIFY(ZMK_KEYMAP_SENSORS_LEN, SENSOR_CHILD_DEFAULTS, (, ), 0)
#endif
};
```

Not sure if you can do math to determine the LISTIFY count, but if so, you could also maybe write it as 

```c
static struct zmk_sensor_config configs[] = {
    DT_FOREACH_CHILD_SEP(ZMK_KEYMAP_SENSORS_NODE, SENSOR_CHILD_ITEM, (, )),
    LISTIFY(ZMK_KEYMAP_SENSORS_LEN - ZMK_KEYMAP_SENSORS_CHILD_COUNT, SENSOR_CHILD_DEFAULTS, (, ), 0)
};
```

Reply:
I prefer the a bit-more explicit first version. Good catch.